File: src/differ_api/api/routes/fs.py
from __future__ import annotations

import os
from pathlib import Path

from fastapi import APIRouter, HTTPException, Query, status
from pydantic import BaseModel

router = APIRouter(prefix="/fs", tags=["filesystem"])


class FsEntry(BaseModel):
    name: str
    path: str
    is_git: bool


class FsBrowseResponse(BaseModel):
    path: str
    parent: str | None
    is_git: bool
    entries: list[FsEntry]


def _looks_like_git_repo(p: Path) -> bool:
    git = p / ".git"
    return git.is_dir() or git.is_file()


def _resolve_start(raw: str | None) -> Path:
    if not raw:
        return Path.home()
    expanded = os.path.expanduser(raw)
    return Path(expanded)


@router.get("/browse", response_model=FsBrowseResponse)
def browse(
    path: str | None = Query(default=None),
    show_hidden: bool = Query(default=False),
) -> FsBrowseResponse:
    target = _resolve_start(path).resolve()
    if not target.exists():
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Path does not exist: {target}",
        )
    if not target.is_dir():
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Path is not a directory: {target}",
        )

    entries: list[FsEntry] = []
    try:
        children = sorted(target.iterdir(), key=lambda p: p.name.lower())
    except PermissionError as err:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=f"Permission denied reading: {target}",
        ) from err

    for child in children:
        if not child.is_dir():
            continue
        if not show_hidden and child.name.startswith("."):
            continue
        entries.append(
            FsEntry(
                name=child.name,
                path=str(child),
                is_git=_looks_like_git_repo(child),
            )
        )

    parent = str(target.parent) if target.parent != target else None
    return FsBrowseResponse(
        path=str(target),
        parent=parent,
        is_git=_looks_like_git_repo(target),
        entries=entries,
    )

File: electron/main/ipc/fs.ts
import { handle } from './utils'
import { readdirSync, statSync } from 'node:fs'
import { join, resolve } from 'node:path'
import { homedir } from 'node:os'

export interface FsEntry {
  name: string
  path: string
  is_git: boolean
}

export interface FsBrowseResponse {
  path: string
  parent: string | null
  is_git: boolean
  entries: FsEntry[]
}

function looksLikeGitRepo(p: string): boolean {
  try {
    const git = join(p, '.git')
    const s = statSync(git)
    return s.isDirectory() || s.isFile()
  } catch {
    return false
  }
}

export function registerFsIpc(): void {
  handle('fs:browse', (payload: unknown) => {
    const req = payload as { path?: string; showHidden?: boolean }
    const target = req.path ? resolve(req.path) : homedir()
    const showHidden = req.showHidden ?? false

    if (!statSync(target).isDirectory()) {
      throw new Error('Path is not a directory')
    }

    const entries: FsEntry[] = []
    const children = readdirSync(target)
      .filter((name) => showHidden || !name.startsWith('.'))
      .map((name) => ({ name, abs: join(target, name) }))
      .filter(({ abs }) => {
        try {
          return statSync(abs).isDirectory()
        } catch {
          return false
        }
      })
      .sort((a, b) => a.name.toLowerCase().localeCompare(b.name.toLowerCase()))

    for (const { name, abs } of children) {
      entries.push({ name, path: abs, is_git: looksLikeGitRepo(abs) })
    }

    const parent = target !== '/' ? resolve(target, '..') : null

    return {
      path: target,
      parent,
      is_git: looksLikeGitRepo(target),
      entries,
    } as FsBrowseResponse
  })
}
